File: experiments/e01-authoring-schema/candidates/c-reflection-go/candidate.go
package candidatec

import (
	"encoding/json"
	"fmt"
	"os"
	"reflect"
	"sort"
	"strconv"
	"strings"
)

type Argument struct {
	Name     string   `json:"name"`
	Type     string   `json:"type"`
	Enum     []string `json:"enum,omitempty"`
	Default  any      `json:"default,omitempty"`
	Required bool     `json:"required"`
}

type Output struct {
	ID       string `json:"id"`
	Type     string `json:"type"`
	Optional bool   `json:"optional"`
}

type Operation struct {
	ID                   string     `json:"id"`
	Description          string     `json:"description"`
	Arguments            []Argument `json:"arguments"`
	Outputs              []Output   `json:"outputs"`
	RequiredEffects      []string   `json:"required_effects"`
	RequiredCapabilities []string   `json:"required_capabilities"`
	body                 any
}

type Schema struct {
	DocumentKind   string      `json:"document_kind"`
	FormatVersion  string      `json:"format_version"`
	FixtureID      string      `json:"fixture_id"`
	FixtureVersion string      `json:"fixture_version"`
	Status         string      `json:"status"`
	Operations     []Operation `json:"operations"`
}

type Diagnostic struct {
	Operation string `json:"operation"`
	Path      string `json:"path"`
	Expected  string `json:"expected"`
	Actual    any    `json:"actual"`
	Human     string `json:"human"`
}

type Handle[T any] struct{ id string }
type Artifact[T any] struct{ Handle[T] }
type Endpoint[T any] struct{ marker func() T }
type Source struct{}
type Check struct{}
type Report[T any] struct{}
type Effect[T any] struct{}
type GoTests struct{}
type Diagnostics struct{}
type BackendBinary struct{}
type IOSApp struct{}
type API struct{}
type OtherAPI struct{}
type MobileE2E struct{}
type PublishedRelease struct{}
type LocalInspection struct{}

type TraceOutput struct {
	ID       string `json:"id"`
	From     string `json:"from,omitempty"`
	Type     string `json:"type"`
	Optional bool   `json:"optional"`
}

type TraceWork struct {
	ID       string        `json:"id"`
	Type     string        `json:"type"`
	Consumes []string      `json:"consumes"`
	Produces []TraceOutput `json:"produces"`
}

type TraceAggregate struct {
	ID        string        `json:"id"`
	Type      string        `json:"type"`
	DependsOn []string      `json:"depends_on"`
	Exposes   []TraceOutput `json:"exposes"`
}

type Relation struct {
	From string `json:"from"`
	To   string `json:"to"`
}

type TraceSource struct {
	ID   string `json:"id"`
	Type string `json:"type"`
}

type Trace struct {
	Status               string         `json:"status"`
	Execution            string         `json:"execution"`
	CompositionRule      string         `json:"composition_rule"`
	Source               TraceSource    `json:"source"`
	ChildWork            []TraceWork    `json:"child_work"`
	Aggregate            TraceAggregate `json:"aggregate"`
	TypedHandleRelations []Relation     `json:"typed_handle_relations"`
}

type traceBuilder struct {
	trace Trace
}

func NewTrace() *traceBuilder {
	return &traceBuilder{trace: Trace{
		Status:          "experiment-only-not-plan-ir",
		Execution:       "fake",
		CompositionRule: "dependencies are authored by passing typed handles, not only string identifiers",
		Source:          TraceSource{ID: "source", Type: "Source"},
	}}
}

func (builder *traceBuilder) Source() Handle[Source] { return Handle[Source]{id: "source"} }

func Child[I, O any](builder *traceBuilder, id string, input Handle[I]) Handle[O] {
	builder.trace.ChildWork = append(builder.trace.ChildWork, TraceWork{
		ID: id, Type: typeName[O](), Consumes: []string{input.id}, Produces: []TraceOutput{},
	})
	builder.trace.TypedHandleRelations = append(builder.trace.TypedHandleRelations, Relation{From: input.id, To: id})
	return Handle[O]{id: id}
}

func Aggregate(builder *traceBuilder, format Handle[Check], tests Handle[Report[GoTests]], lint Handle[Check]) Handle[Check] {
	for _, input := range []string{format.id, tests.id, lint.id} {
		builder.trace.TypedHandleRelations = append(builder.trace.TypedHandleRelations, Relation{From: input, To: "check"})
	}
	builder.trace.Aggregate = TraceAggregate{ID: "check", Type: "Check", DependsOn: []string{format.id, tests.id, lint.id}}
	return Handle[Check]{id: "check"}
}

func RequiredOutput[T any](id string, from Handle[T]) TraceOutput {
	return TraceOutput{ID: id, From: from.id, Type: typeName[T]()}
}

func OptionalOutput[T any](id string, from Handle[Check]) TraceOutput {
	return TraceOutput{ID: id, From: from.id, Type: typeName[T](), Optional: true}
}

func (builder *traceBuilder) Finish(outputs ...TraceOutput) Trace {
	builder.trace.Aggregate.Exposes = outputs
	for _, output := range outputs {
		if output.From == "test" {
			for index := range builder.trace.ChildWork {
				if builder.trace.ChildWork[index].ID == "test" {
					produced := output
					produced.From = ""
					builder.trace.ChildWork[index].Produces = []TraceOutput{produced}
				}
			}
		}
	}
	return builder.trace
}

type definition struct {
	Scope                string
	FixtureID            string
	FixtureVersion       string
	Status               string
	OperationID          string
	Description          string
	RequiredEffects      []string
	RequiredCapabilities []string
	Body                 any
}

// E01-AUTHOR-BEGIN
func ComposeW1() Trace {
	trace := NewTrace()
	source := trace.Source()
	format := Child[Source, Check](trace, "format", source)
	tests := Child[Source, Report[GoTests]](trace, "test", source)
	lint := Child[Source, Check](trace, "lint", source)
	check := Aggregate(trace, format, tests, lint)
	return trace.Finish(
		RequiredOutput("test-report", tests),
		OptionalOutput[Report[Diagnostics]]("diagnostics", check),
	)
}

type W1Arguments struct {
	Verbosity   string `e01:"name=verbosity,enum=quiet|normal|verbose,default=normal"`
	ChangedOnly bool   `e01:"name=changed-only,default=false"`
}

type W1Outputs struct {
	TestReport  Report[GoTests]     `e01:"id=test-report"`
	Diagnostics Report[Diagnostics] `e01:"id=diagnostics,optional=true"`
}

var w1 = definition{
	Scope:                "W1",
	FixtureID:            "w1-fast-project-check",
	FixtureVersion:       "t1-experimental-v1",
	Status:               "experimental",
	OperationID:          "check",
	Description:          "Format check, unit tests, and static analysis, aggregated into one pass/fail Check.",
	RequiredCapabilities: []string{"filesystem-read"},
	Body:                 neverRun[W1Arguments, W1Outputs]("W1"),
}

// E01-AUTHOR-END

type NoArguments struct{}
type W2Outputs struct {
	Backend    Artifact[BackendBinary] `e01:"id=backend-binary"`
	Tests      Report[GoTests]         `e01:"id=go-tests-report"`
	Inspection LocalInspection         `e01:"id=inspection-summary"`
}
type W3Outputs struct {
	Report Report[MobileE2E] `e01:"id=mobile-e2e-report"`
}
type EffectArguments struct {
	Environment string `e01:"name=environment,enum=staging|production,required=true"`
	Channel     string `e01:"name=channel,enum=beta|stable,default=beta"`
}
type EffectOutputs struct {
	Published Effect[PublishedRelease] `e01:"id=published-release"`
}

func definitions() map[string]definition {
	return map[string]definition{
		"W1": w1,
		"W2": {
			Scope: "W2", FixtureID: "w2-cross-target-artifact-pipeline", FixtureVersion: "t1-w2-experimental-v1", Status: "experimental",
			OperationID: "build-and-verify", Description: "Build the backend binary on Linux, run its Go test suite on a compatible Linux worker, and produce a local inspection summary.",
			RequiredCapabilities: []string{"linux-execution-profile"}, Body: neverRun[NoArguments, W2Outputs]("W2"),
		},
		"W3": {
			Scope: "W3", FixtureID: "w3-isolated-native-mobile-stack", FixtureVersion: "t1-w3-fixture-v1-experimental", Status: "experimental",
			OperationID: "mobile-e2e", Description: "Bring up a namespace-private Linux API stack, build the iOS app on macOS/Xcode, and run the end-to-end suite against a simulator.",
			RequiredCapabilities: []string{"linux-execution-profile", "macos-execution-profile", "simulator-session"}, Body: neverRun[NoArguments, W3Outputs]("W3"),
		},
		"effect": {
			Scope: "effect", FixtureID: "e01-effect-probe", FixtureVersion: "e01-effect-probe-v1", Status: "experimental-synthetic",
			OperationID: "publish-preview", Description: "Describe a policy-gated release publication without evaluating or authorizing its operation body.",
			RequiredEffects: []string{"publish-release"}, RequiredCapabilities: []string{"network:app-store-connect", "secret:app-store-signing"},
			Body: neverRun[EffectArguments, EffectOutputs]("effect"),
		},
	}
}

func neverRun[A, O any](scope string) func(A) O {
	return func(A) O {
		if path := os.Getenv("E01_BODY_SENTINEL"); path != "" {
			_ = os.WriteFile(path, []byte(scope), 0o600)
		}
		panic("E01 operation body evaluated during discovery: " + scope)
	}
}

func Discover(scope string) (Schema, error) {
	item, ok := definitions()[scope]
	if !ok {
		return Schema{}, fmt.Errorf("unknown scope %q", scope)
	}
	function := reflect.TypeOf(item.Body)
	if function.Kind() != reflect.Func || function.NumIn() != 1 || function.NumOut() != 1 {
		return Schema{}, fmt.Errorf("operation %s must have one argument and one output struct", item.OperationID)
	}
	arguments, err := reflectArguments(function.In(0))
	if err != nil {
		return Schema{}, fmt.Errorf("operation %s arguments: %w", item.OperationID, err)
	}
	outputs, err := reflectOutputs(function.Out(0))
	if err != nil {
		return Schema{}, fmt.Errorf("operation %s outputs: %w", item.OperationID, err)
	}
	operation := Operation{
		ID: item.OperationID, Description: item.Description, Arguments: arguments, Outputs: outputs,
		RequiredEffects: nonNil(item.RequiredEffects), RequiredCapabilities: nonNil(item.RequiredCapabilities), body: item.Body,
	}
	return Schema{
		DocumentKind: "schema", FormatVersion: "t1-plan-conformance-schema-v1", FixtureID: item.FixtureID,
		FixtureVersion: item.FixtureVersion, Status: item.Status, Operations: []Operation{operation},
	}, nil
}

func reflectArguments(value reflect.Type) ([]Argument, error) {
	if value.Kind() != reflect.Struct {
		return nil, fmt.Errorf("%s is not an argument struct", value)
	}
	arguments := make([]Argument, 0, value.NumField())
	for index := 0; index < value.NumField(); index++ {
		field := value.Field(index)
		options := parseTag(field.Tag.Get("e01"))
		name := options["name"]
		if name == "" {
			return nil, fmt.Errorf("%s.%s missing e01 name", value.Name(), field.Name)
		}
		argument := Argument{Name: name, Type: scalarName(field.Type)}
		if members := options["enum"]; members != "" {
			argument.Enum = strings.Split(members, "|")
		}
		if raw, found := options["default"]; found {
			argument.Default = parseDefault(raw, field.Type)
		}
		argument.Required = options["required"] == "true"
		arguments = append(arguments, argument)
	}
	return arguments, nil
}

func reflectOutputs(value reflect.Type) ([]Output, error) {
	if value.Kind() != reflect.Struct {
		return nil, fmt.Errorf("%s is not an output struct", value)
	}
	outputs := make([]Output, 0, value.NumField())
	for index := 0; index < value.NumField(); index++ {
		field := value.Field(index)
		options := parseTag(field.Tag.Get("e01"))
		id := options["id"]
		if id == "" {
			return nil, fmt.Errorf("%s.%s missing e01 id", value.Name(), field.Name)
		}
		outputs = append(outputs, Output{ID: id, Type: reflectedTypeName(field.Type), Optional: options["optional"] == "true"})
	}
	return outputs, nil
}

func parseTag(raw string) map[string]string {
	values := map[string]string{}
	for _, part := range strings.Split(raw, ",") {
		if part == "" {
			continue
		}
		pieces := strings.SplitN(part, "=", 2)
		if len(pieces) == 2 {
			values[pieces[0]] = pieces[1]
		}
	}
	return values
}

func parseDefault(raw string, value reflect.Type) any {
	if value.Kind() == reflect.Bool {
		parsed, _ := strconv.ParseBool(raw)
		return parsed
	}
	return raw
}

func scalarName(value reflect.Type) string {
	switch value.Kind() {
	case reflect.String:
		return "string"
	case reflect.Bool:
		return "boolean"
	default:
		return value.Name()
	}
}

func reflectedTypeName(value reflect.Type) string {
	name := value.Name()
	name = strings.ReplaceAll(name, "e01/candidatec.", "")
	return name
}

func typeName[T any]() string {
	return reflectedTypeName(reflect.TypeOf((*T)(nil)).Elem())
}

func nonNil(values []string) []string {
	if values == nil {
		return []string{}
	}
	return values
}

func Validate(scope string, values map[string]any) *Diagnostic {
	schema, err := Discover(scope)
	if err != nil {
		return &Diagnostic{Operation: scope, Path: "$", Expected: "known operation", Actual: scope, Human: err.Error()}
	}
	operation := schema.Operations[0]
	known := map[string]Argument{}
	for _, argument := range operation.Arguments {
		known[argument.Name] = argument
	}
	keys := make([]string, 0, len(values))
	for key := range values {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		if _, found := known[key]; !found {
			return diagnostic(operation.ID, key, "known argument", values[key])
		}
	}
	for _, argument := range operation.Arguments {
		value, present := values[argument.Name]
		if !present {
			if argument.Required {
				return diagnostic(operation.ID, argument.Name, "required "+argument.Type, nil)
			}
			continue
		}
		if argument.Type == "string" {
			if _, ok := value.(string); !ok {
				return diagnostic(operation.ID, argument.Name, "string", value)
			}
		}
		if argument.Type == "boolean" {
			if _, ok := value.(bool); !ok {
				return diagnostic(operation.ID, argument.Name, "boolean", value)
			}
		}
		if len(argument.Enum) > 0 {
			text, ok := value.(string)
			if ok && !contains(argument.Enum, text) {
				return diagnostic(operation.ID, argument.Name, "one of "+fmt.Sprint(argument.Enum), value)
			}
		}
	}
	return nil
}

func contains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

func diagnostic(operation, path, expected string, actual any) *Diagnostic {
	return &Diagnostic{
		Operation: operation, Path: path, Expected: expected, Actual: actual,
		Human: fmt.Sprintf("operation %s argument %s expected %s; got %v", operation, path, expected, actual),
	}
}

func Encode(value any) ([]byte, error)         { return json.MarshalIndent(value, "", "  ") }
func AcceptIOS(artifact Artifact[IOSApp])      { _ = artifact }
func ConnectOther(endpoint Endpoint[OtherAPI]) { _ = endpoint }

File: experiments/e01-authoring-schema/candidates/d-typescript/candidate.ts
import { writeFileSync } from "node:fs";

export type Argument = {
  name: string;
  type: string;
  enum?: string[];
  default?: unknown;
  required: boolean;
};

export type Output = {
  id: string;
  type: string;
  optional: boolean;
};

type Operation = {
  id: string;
  description: string;
  arguments: Argument[];
  outputs: Output[];
  required_effects: string[];
  required_capabilities: string[];
  body: () => never;
};

export type Schema = {
  document_kind: "schema";
  format_version: "t1-plan-conformance-schema-v1";
  fixture_id: string;
  fixture_version: string;
  status: string;
  operations: Omit<Operation, "body">[];
};

export type Diagnostic = {
  operation: string;
  path: string;
  expected: string;
  actual: unknown;
  human: string;
};

export interface BackendBinary {
  readonly __backendBinary: unique symbol;
}
export interface IOSApp {
  readonly __iosApp: unique symbol;
}
export interface API {
  readonly __api: unique symbol;
}
export interface OtherAPI {
  readonly __otherAPI: unique symbol;
}

export class Artifact<T> {
  private invariant!: (value: T) => T;
}

export class Endpoint<T> {
  private invariant!: (value: T) => T;
}

class Handle<T> {
  private invariant!: (value: T) => T;
  constructor(readonly id: string) {}
}

type Source = { readonly source: unique symbol };
type Check = { readonly check: unique symbol };
type GoTests = { readonly goTests: unique symbol };
type Report<T> = { readonly report: T };
type Diagnostics = { readonly diagnostics: unique symbol };

type TraceOutput = {
  id: string;
  from?: string;
  type: string;
  optional: boolean;
};

type TraceWork = {
  id: string;
  type: string;
  consumes: string[];
  produces: TraceOutput[];
};

export type Trace = {
  status: "experiment-only-not-plan-ir";
  execution: "fake";
  composition_rule: string;
  source: { id: string; type: "Source" };
  child_work: TraceWork[];
  aggregate: {
    id: string;
    type: "Check";
    depends_on: string[];
    exposes: TraceOutput[];
  };
  typed_handle_relations: { from: string; to: string }[];
};

class TraceBuilder {
  private readonly children: TraceWork[] = [];
  private readonly relations: { from: string; to: string }[] = [];

  source(): Handle<Source> {
    return new Handle<Source>("source");
  }

  child<I, O>(id: string, type: string, input: Handle<I>): Handle<O> {
    this.children.push({ id, type, consumes: [input.id], produces: [] });
    this.relations.push({ from: input.id, to: id });
    return new Handle<O>(id);
  }

  aggregate(
    format: Handle<Check>,
    tests: Handle<Report<GoTests>>,
    lint: Handle<Check>,
  ): Handle<Check> {
    for (const input of [format, tests, lint]) {
      this.relations.push({ from: input.id, to: "check" });
    }
    return new Handle<Check>("check");
  }

  finish(testReport: TraceOutput, diagnostics: TraceOutput): Trace {
    const test = this.children.find((child) => child.id === "test");
    if (test === undefined) {
      throw new Error("W1 trace requires test child");
    }
    test.produces = [{ ...testReport, from: undefined }];
    return {
      status: "experiment-only-not-plan-ir",
      execution: "fake",
      composition_rule: "dependencies are authored by passing typed handles, not only string identifiers",
      source: { id: "source", type: "Source" },
      child_work: this.children,
      aggregate: {
        id: "check",
        type: "Check",
        depends_on: ["format", "test", "lint"],
        exposes: [testReport, diagnostics],
      },
      typed_handle_relations: this.relations,
    };
  }
}

function requiredOutput<T>(id: string, from: Handle<T>, type: string): TraceOutput {
  return { id, from: from.id, type, optional: false };
}

function optionalOutput<T>(id: string, from: Handle<T>, type: string): TraceOutput {
  return { id, from: from.id, type, optional: true };
}

function neverRun(scope: string): () => never {
  return () => {
    const path = process.env.E01_BODY_SENTINEL;
    if (path !== undefined) {
      writeFileSync(path, scope, { mode: 0o600 });
    }
    throw new Error(`E01 operation body evaluated during discovery: ${scope}`);
  };
}

function operation(
  fixtureId: string,
  fixtureVersion: string,
  status: string,
  value: Operation,
): Schema {
  const { body: _body, ...publicValue } = value;
  return {
    document_kind: "schema",
    format_version: "t1-plan-conformance-schema-v1",
    fixture_id: fixtureId,
    fixture_version: fixtureVersion,
    status,
    operations: [publicValue],
  };
}

// E01-AUTHOR-BEGIN
export function composeW1(): Trace {
  const trace = new TraceBuilder();
  const source = trace.source();
  const format = trace.child<Source, Check>("format", "Check", source);
  const tests = trace.child<Source, Report<GoTests>>("test", "Report[GoTests]", source);
  const lint = trace.child<Source, Check>("lint", "Check", source);
  const check = trace.aggregate(format, tests, lint);
  return trace.finish(
    requiredOutput("test-report", tests, "Report[GoTests]"),
    optionalOutput("diagnostics", check, "Report[Diagnostics]"),
  );
}

const w1 = operation(
  "w1-fast-project-check",
  "t1-experimental-v1",
  "experimental",
  {
    id: "check",
    description: "Format check, unit tests, and static analysis, aggregated into one pass/fail Check.",
    arguments: [
      {
        name: "verbosity",
        type: "string",
        enum: ["quiet", "normal", "verbose"],
        default: "normal",
        required: false,
      },
      {
        name: "changed-only",
        type: "boolean",
        default: false,
        required: false,
      },
    ],
    outputs: [
      { id: "test-report", type: "Report[GoTests]", optional: false },
      { id: "diagnostics", type: "Report[Diagnostics]", optional: true },
    ],
    required_effects: [],
    required_capabilities: ["filesystem-read"],
    body: neverRun("W1"),
  },
);
// E01-AUTHOR-END

const w2 = operation(
  "w2-cross-target-artifact-pipeline",
  "t1-w2-experimental-v1",
  "experimental",
  {
    id: "build-and-verify",
    description: "Build the backend binary on Linux, run its Go test suite on a compatible Linux worker, and produce a local inspection summary.",
    arguments: [],
    outputs: [
      { id: "backend-binary", type: "Artifact[BackendBinary]", optional: false },
      { id: "go-tests-report", type: "Report[GoTests]", optional: false },
      { id: "inspection-summary", type: "LocalInspection", optional: false },
    ],
    required_effects: [],
    required_capabilities: ["linux-execution-profile"],
    body: neverRun("W2"),
  },
);

const w3 = operation(
  "w3-isolated-native-mobile-stack",
  "t1-w3-fixture-v1-experimental",
  "experimental",
  {
    id: "mobile-e2e",
    description: "Bring up a namespace-private Linux API stack, build the iOS app on macOS/Xcode, and run the end-to-end suite against a simulator.",
    arguments: [],
    outputs: [{ id: "mobile-e2e-report", type: "Report[MobileE2E]", optional: false }],
    required_effects: [],
    required_capabilities: [
      "linux-execution-profile",
      "macos-execution-profile",
      "simulator-session",
    ],
    body: neverRun("W3"),
  },
);

const effect = operation(
  "e01-effect-probe",
  "e01-effect-probe-v1",
  "experimental-synthetic",
  {
    id: "publish-preview",
    description: "Describe a policy-gated release publication without evaluating or authorizing its operation body.",
    arguments: [
      {
        name: "environment",
        type: "string",
        enum: ["staging", "production"],
        required: true,
      },
      {
        name: "channel",
        type: "string",
        enum: ["beta", "stable"],
        default: "beta",
        required: false,
      },
    ],
    outputs: [{ id: "published-release", type: "Effect[PublishedRelease]", optional: false }],
    required_effects: ["publish-release"],
    required_capabilities: ["network:app-store-connect", "secret:app-store-signing"],
    body: neverRun("effect"),
  },
);

const schemas: Record<string, Schema> = { W1: w1, W2: w2, W3: w3, effect };

export function discover(scope: string): Schema {
  const schema = schemas[scope];
  if (schema === undefined) {
    throw new Error(`unknown scope ${JSON.stringify(scope)}`);
  }
  return schema;
}

export function validate(scope: string, values: Record<string, unknown>): Diagnostic | undefined {
  const schema = discover(scope);
  const candidate = schema.operations[0];
  if (candidate === undefined) {
    throw new Error(`scope ${scope} has no operation`);
  }
  const known = new Map(candidate.arguments.map((argument) => [argument.name, argument]));
  for (const key of Object.keys(values).sort()) {
    if (!known.has(key)) {
      return diagnostic(candidate.id, key, "known argument", values[key]);
    }
  }
  for (const argument of candidate.arguments) {
    const present = Object.prototype.hasOwnProperty.call(values, argument.name);
    const value = values[argument.name];
    if (!present) {
      if (argument.required) {
        return diagnostic(candidate.id, argument.name, `required ${argument.type}`, null);
      }
      continue;
    }
    if (argument.type === "string" && typeof value !== "string") {
      return diagnostic(candidate.id, argument.name, "string", value);
    }
    if (argument.type === "boolean" && typeof value !== "boolean") {
      return diagnostic(candidate.id, argument.name, "boolean", value);
    }
    if (argument.enum !== undefined && typeof value === "string" && !argument.enum.includes(value)) {
      return diagnostic(candidate.id, argument.name, `one of [${argument.enum.join(" ")}]`, value);
    }
  }
  return undefined;
}

function diagnostic(operationId: string, path: string, expected: string, actual: unknown): Diagnostic {
  return {
    operation: operationId,
    path,
    expected,
    actual,
    human: `operation ${operationId} argument ${path} expected ${expected}; got ${String(actual)}`,
  };
}

export function encode(value: unknown): string {
  return `${JSON.stringify(value, null, 2)}\n`;
}

export function acceptIOS(_artifact: Artifact<IOSApp>): void {}
export function connectOther(_endpoint: Endpoint<OtherAPI>): void {}
